File: array/48-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        int l = matrix.length;

        /* Perform Tranposition */
        for (int i = 0; i < l; i++) {
            for (int j = i; j < l; j++){
                int cache = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = cache;
            }
        }
        
        /* Perform Horizontal Reversal */
        for (int i = 0; i < l; i++) {
            for (int j = 0; j < l / 2; j++){
                int cache = matrix[i][j];
                matrix[i][j] = matrix[i][l - 1 - j];
                matrix[i][l - 1 - j] = cache;
            }
        }
    }
}

File: array/48-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        l = len(matrix)
        
        # Perform Transpose
        for i in range(l):
            for j in range(i, l):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        
        print(matrix)
        
        # Perform reversal
        for i in range(l):
            for j in range(l // 2):
                matrix[i][j], matrix[i][l - 1 - j] = matrix[i][l - 1 - j], matrix[i][j]